plugins/org.polarsys.capella.scenario.editor/src/org/polarsys/capella/scenario/editor/helper/EmbeddedEditorInstanceHelper.java

```diff
@@ -297,7 +300,91 @@ public static boolean checkValidKeyword(String keyword) {
     }
     return false;
   }
-  
+
+  public static boolean checkValidTimeline(String timelineName) {
+    if (timelineName == null || getElementWithGivenName(timelineName) == null) 
+      return false;
+    return true;
+  }
+
+  private static EObject getElementWithGivenName(String timelineName) {
+    List<String> keywords = Arrays.asList(DslConstants.ACTOR, DslConstants.FUNCTION, DslConstants.ACTIVITY,
+        DslConstants.COMPONENT, DslConstants.ROLE);
+    for (String keyword : keywords) {   
+      if (checkValidKeyword(keyword)) {     
+        Collection<? extends EObject> availableElements = getAvailableElements(keyword);
+        for (EObject element : availableElements) {
+          String elementName = ((AbstractInstance) element).getName();
+
+          if (elementName.equals(timelineName)) {
+            return element;
+          }
+        }
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Get all available state fragments (mode, state or allocated function) for a specific instance role
+   * 
+   * @param type
+   *          - the type of state fragment: mode, state or allocated function
+   * @param timelineName
+   *          - the timeline name
+   * @return List<String> - return the list with all available state fragments
+   */
+  public static List<String> getAvailableStateFragments(String type, String timelineName) {
+    if (timelineName == null || type == null) {
+      return new ArrayList<String>();
+    }
+
+    EObject element = getElementWithGivenName(timelineName);
+    if (element != null) {
+        if (type.equals(DslConstants.FUNCTION)) {
+          return getAllocatedFunctionsName(element);
+        }
+        return getStateOrModeNames(element, type);
+      }
```

Comment:
can remove this function, getStateOrModeNames an instead add two calls one to getStateNames one to getModeNames.